packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -321,69 +491,55 @@ function* trackObjectNodeDiscrepancies(
 	 * Then, the stored schema is iterated to find the third type.
 	 */
 
-	for (const result of compareMaps(view.objectNodeFields, stored.objectNodeFields)) {
-		const fieldKey = result.key;
-		switch (result.type) {
-			case "aExtra": {
-				if (result.value.kind === storedEmptyFieldSchema.kind) {
-					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
-					// (by way of omission). We treat these identically anyway.
-					break;
-				}
-				yield {
-					identifier,
-					fieldKey,
-					mismatch: "fieldKind",
-					view: result.value.kind,
-					stored: storedEmptyFieldSchema.kind,
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "bExtra": {
-				if (result.value.kind === storedEmptyFieldSchema.kind) {
-					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
-					// (by way of omission). We treat these identically anyway.
-					break;
-				}
-				yield {
-					identifier,
-					fieldKey,
-					mismatch: "fieldKind",
-					view: storedEmptyFieldSchema.kind,
-					stored: result.value.kind,
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "both": {
-				yield* getFieldDiscrepancies(result.valueA, result.valueB, identifier, fieldKey);
-				break;
-			}
-			default:
-				break;
+	const storedToSimpleKeys = new Map<FieldKey, string | symbol>();
+
+	for (const [key, { storedKey, schema }] of view) {
+		const storedSchema = stored.objectNodeFields.get(storedKey);
+		storedToSimpleKeys.set(storedKey, key);
+
+		// If the view schema has a field that's not in the stored schema
+		if (storedSchema === undefined) {
+			const viewKind =
+				convertFieldKind.get(schema.kind) ??
+				fail("A conversion from a FieldKind to a FlexFieldKind should exist");
+			yield {
+				identifier,
+				fieldKey: storedKey,
+				mismatch: "fieldKind",
+				view: viewKind.identifier,
+				stored: storedEmptyFieldSchema.kind,
+			} satisfies FieldKindDiscrepancy;
+		} else {
+			yield* getFieldDiscrepancies(
+				schema,
+				storedSchema,
+				identifier,
+				storedKey,
+				viewKindIsSequence,
+			);
 		}
 	}
-}
 
-function* compareMaps<K, V1, V2>(
-	a: ReadonlyMap<K, V1>,
-	b: ReadonlyMap<K, V2>,
-): Iterable<
-	| { type: "aExtra"; key: K; value: V1 }
-	| { type: "bExtra"; key: K; value: V2 }
-	| { type: "both"; key: K; valueA: V1; valueB: V2 }
-> {
-	for (const [key, valueA] of a) {
-		const valueB = b.get(key);
-		yield valueB === undefined
-			? { type: "aExtra", key, value: valueA }
-			: { type: "both", key, valueA, valueB };
-	}
-	for (const [key, valueB] of b) {
-		if (!a.has(key)) {
-			yield { type: "bExtra", key, value: valueB };
+	for (const [fieldKey, schema] of stored.objectNodeFields) {
+		const viewSchema = storedToSimpleKeys.get(fieldKey);
+		// If the stored schema has a field that's not in the view schema
+		if (viewSchema === undefined) {
```

Comment:
This is a bit confusing. We create the storedToSimpleKeys, but actually only use it as a set of FieldKeys used by the viewSchema.

Also calling the propertyKey viewSchema is confusing, since its a key not a schema.

Replacing `storedToSimpleKeys` with a `Set<FieldKey>` called `viewFields ` then doing

```suggestion
		// If the stored schema has a field that's not in the view schema
		if (!viewFields.has(fieldKey)) {
```

could be simpler and more clear